plumbing/format/packfile/decoder.go

```diff
@@ -174,24 +185,54 @@ func (d *Decoder) decodeObjectsWithObjectStorerTx(count int) error {
 
 // DecodeObject reads the next object from the scanner and returns it. This
 // method can be used in replacement of the Decode method, to work in a
-// interative way
+// interactive way. If you created a new decoder instance using NewDecoderForType
+// constructor, if the object decoded is not equals to the specified one, null will
+// be returned
 func (d *Decoder) DecodeObject() (plumbing.EncodedObject, error) {
+	h, err := d.nextHeader()
+	if err != nil {
+		return nil, err
+	}
+
+	var realType plumbing.ObjectType
+	switch {
```

Comment:
change this with `switch h.Type`, then you can use the values directly in cases.